semgrep-core/src/tainting/Dataflow_tainting.ml

```diff
@@ -557,34 +641,52 @@ let input_env ~enter_env ~(flow : F.cfg) mapping ni =
 let (transfer :
       config ->
       fun_env ->
-      Taints.t Dataflow_core.env ->
+      taint_info Dataflow_core.env ->
       string option ->
       flow:F.cfg ->
-      Taints.t Dataflow_core.transfn) =
+      taint_info Dataflow_core.transfn) =
  fun config fun_env enter_env opt_name ~flow
      (* the transfer function to update the mapping at node index ni *)
        mapping ni ->
   (* DataflowX.display_mapping flow mapping show_tainted; *)
-  let in' : Taints.t VarMap.t = input_env ~enter_env ~flow mapping ni in
+  let in' : taint_info VarMap.t = input_env ~enter_env ~flow mapping ni in
   let node = flow.graph#nodes#assoc ni in
-  let out' : Taints.t VarMap.t =
-    let env = { config; fun_name = opt_name; fun_env; var_env = in' } in
+  let out' : taint_info VarMap.t =
+    let env =
+      {
+        config;
+        fun_name = opt_name;
+        fun_env;
+        var_env = in';
+        field_env = Hashtbl.create 100;
+      }
+    in
     match node.F.n with
     | NInstr x -> (
         let taints, var_env' = check_tainted_instr env x in
         let var_env' =
           match LV.lvar_of_instr_opt x with
           | None -> var_env'
-          | Some var ->
+          | Some (_, var, _) ->
               (* We call `check_tainted_var` here because the assigned `var`
                * itself could be annotated as a source of taint. *)
               check_tainted_var { env with var_env = var_env' } var |> snd
         in
         match (Taints.is_empty taints, LV.lvar_of_instr_opt x) with
         (* Instruction returns safe data, remove taint from `var`. *)
-        | true, Some var -> VarMap.remove (str_of_name var) var_env'
+        | true, Some (_, var, _) ->
+            let var = str_of_name var in
+            VarMap.fold
+              (fun str taint map ->
+                if is_dotted_prefix var str then VarMap.add str MarkedClean map
+                else VarMap.add str taint map)
+              (VarMap.add var MarkedClean var_env')
+              VarMap.empty
         (* Instruction returns tainted data, add taints to `var`. *)
-        | false, Some var -> add_taint_to_var_in_env var_env' var taints
+        | false, Some (base_name, _, dots) ->
+            List.fold_left
+              (fun var_env var -> add_taint_to_var_in_env var_env var taints)
```

Comment:
As I said, I think in this case we should just taint the `last dots` or `base_name` if `dots = []`. But given `a.b.c = user_input` I don't think we should be tainting `a` and `a.b`, just `a.b.c`. Due to aliasing, it could happen that tainting `a.b.c` should taint `a` itself, but I think we don't need to handle those cases, better to report fewer FPs.